core/src/main/java/io/grpc/DnsNameResolver.java

```diff
@@ -0,0 +1,116 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+
+/**
+ * A DNS-based {@link NameResolver}.
+ */
+@ExperimentalApi
+public final class DnsNameResolver implements NameResolver {
+
+  private static final DnsNameResolver instance = new DnsNameResolver();
+
+  @Override
+  public Handle resolve(String scheme, @Nullable String nsAuthority, String endpoint) {
+    return new HandleImpl(nsAuthority, endpoint);
+  }
+
+  private DnsNameResolver() {
+  }
+
+  public static DnsNameResolver getInstance() {
+    return instance;
+  }
+
+  private static class HandleImpl implements Handle {
+    private final String authority;
+    private final String host;
+    private final int port;
+    private Executor executor;
+
+    HandleImpl(@Nullable String nsAuthority, String endpoint) {
+      // TODO: if a DNS server is provided as target.getAuthority(), use it.
+      // https://www.captechconsulting.com/blogs/accessing-the-dusty-corners-of-dns-with-java
+
+      // Must prepend a "//" to the endpoint when constructing a URI, otherwise
+      // the authority and host of the resulted URI would be null.
+      URI endpointUri = URI.create("//" + endpoint);
+      authority = Preconditions.checkNotNull(endpointUri.getAuthority(),
+          "endpointUri (%s) doesn't have an authority", endpointUri);
+      host = Preconditions.checkNotNull(endpointUri.getHost(), "host");
+      port = endpointUri.getPort();
+      Preconditions.checkArgument(port > 0, "port (%s) must be positive", port);
+    }
+
+    @Override
+    public String getServiceAuthority() {
+      return authority;
+    }
+
+    @Override
+    public synchronized void start(Executor executor, final Listener listener) {
+      Preconditions.checkState(this.executor == null, "Already started");
+      this.executor = executor;
+      executor.execute(new Runnable() {
```

Comment:
Done. Also documented that methods of `NameResolver.Listener` and `LoadBalancer` should not block.